### DISCUSSION TITLE
Local Only Mode

### DISCUSSION THREAD

**Discussion by MentalGear** [electric-sql/electric#883] [Feature requests]
First of all: Very impressive work, looking forward to add this into our techstack. Thank you!

I'm currently building a small browser extension and, before I add any remote server sync/storage, I'd like to start small and only save data locally. Is it possible to use electric without a server pg backend  - so it runs completely local ?

PS: would it also be possible to use pgvector with electric?

**Comment by kevin-dp** [comment]
Hi @MentalGear,
It is perfectly possible to run an Electric client without the Electric backend.
In that case the client will run in offline mode so it won't be able to connect to the backend but will still be fully functional meaning that you can read and write the local SQLite DB.

I will let @thruflo answer your pgvector question as he has a better viewpoint on our roadmap.

**Comment by thruflo** [comment]
Hey, how would you like to use pgvector?

**Comment by MentalGear** [comment]
Thanks for getting back, I appreciate it!

**Local-Mode**
Would it be possible to outline the setup process of having electric run in "local-mode"? As I understand, normally one needs to have a postgres back-end in place to "electrify" tables - or can it be done without? Would I need to go through the normal setup and then just disconnect the database in code?

As I think about it, it'd be great if you had a page on the docs for this kind of "local-mode starter setup".
Lots of apps start small and only need local storage at first: it would be a _huge USP_ if a project can start out with electric in the initial stack for simple client-side storage and, once more advanced sync features are needed, seamlessly mature into using a backend db without rewriting the entire codebase - just by connecting the database!

**Local Storage Considerations**
Are there any risks of running in local-mode for a prolonged amount of time? (ex: time/space limit). It would be a huge frustration for users if the app would just lose data after x days because of some cache cleaning/storage limitation.

**pg_vector**
I'd like to use a vector search for clustering similar textual data client-side. I guess vector search can't be done directly on electric, therefore I have been exploring using client-side [postgres-wasm](https://github.com/snaplet/postgres-wasm) or "voy-search".
I guess if using postgres-wasm, electric can return vector search results like normal queries ?

**Comment by kevin-dp** [comment]
> Would it be possible to outline the setup process of having electric run in "local-mode"? As I understand, normally one needs to have a postgres back-end in place to "electrify" tables - or can it be done without? Would I need to go through the normal setup and then just disconnect the database in code?

The app will run in local-mode if it fails to reach Electric (i.e. the server).
To setup the app you can indeed go through the regular setup process in order to generate the client and then not run the Electric backend.

> Are there any risks of running in local-mode for a prolonged amount of time? (ex: time/space limit). It would be a huge frustration for users if the app would just lose data after x days because of some cache cleaning/storage limitation.

SQLite stores the DB as a single file on disk, so you are bound by the maximum file size.
SQLite has a great page that documents various limitations: [https://www.sqlite.org/whentouse.html](https://www.sqlite.org/whentouse.html).

For Electric apps running in the browser, we currently use wa-sqlite and configured it to use IndexedDB as the virtual file system. Storage limits for IndexedDB are browser-specific and are described here: [https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#Storage_limits](https://developer.mozilla.org/en-US/docs/Web/API/Storage_API/Storage_quotas_and_eviction_criteria#Storage_limits)

**Comment by thruflo** [comment]
> pg_vector
> I'd like to use a vector search for clustering similar textual data client-side. I guess vector search can't be done directly on electric, therefore I have been exploring using client-side [postgres-wasm](https://github.com/snaplet/postgres-wasm) or "voy-search".
> I guess if using postgres-wasm, electric can return vector search results like normal queries ?

There's some devil in the detail here relating to type mapping, autogenerated queries etc. So you can't just swap in a VM-based WASM PG. However, we have been working on this, or more specifically @balegas and @tachyonicbytes have been working on this, in a slightly different configuration.

We have a public #wasm-and-rust channel on the Discord if you'd like to say hi there / discuss more.

**Comment by jtkeyva** [comment]
any movement on this? thanks

**Reply by balegas** [reply]
Hi, sorry for the late response.

It is possible for a while to use Electric without a sync service https://electric-sql.com/docs/api/cli#local-only-first-mode

**Comment by fredericrous** [comment]
> The app will run in local-mode if it fails to reach Electric (i.e. the server).
> To setup the app you can indeed go through the regular setup process in order to generate the client and then not run the Electric backend.

a precision tho, `electric.connect` method will throw. In the example generated with create-electric-app, we have to comment `await electric.connect(authToken())`

**Reply by balegas** [reply]
Yeah, indeed that was one of the changes we made to support a local-only mode. Before, electrification would imply trying to connect to the sync service.

**Comment by KyleAMathews** [comment]
👋 we've been working the last month on a rebuild of the Electric server over at a temporary repo https://github.com/electric-sql/electric-next/

You can read more about why we made the decision at https://next.electric-sql.com/about

We're really excited about all the new possibilities the new server brings and we hope you'll check it out soon and give us your feedback.

We're now moving the temporary repo back here. As part of that migration we're closing all the old issues, PRs, and discussions. We really appreciate you taking the time to start this discussion! If it's still relevant in the new system, please reopen.
